[ota-requestor] Switch shell commands to use the common code

Damian-Nordic opened the pull request:
#### Problem
OTA shell commands contain a custom implementation (that is, another copy-paste) for connecting and communicating with an OTA provider instead of using `OTARequestor` class.

#### Change overview
* Rewrite the commands to use the `OTARequestor` class to call an OTA provider.
* Provide `OTAImageProcessorImpl` classes for nRF Connect and ESP32.
* Enable building `src/clusters/ota-requestor` when `OTA Provider client` is enabled in ZAP.

#### Testing
Tested using nRF Connect lighting-app and Linux OTA Provider

**NOTE: Three commits are shared with #12796 so make sure to disable those commits when reviewing this PR**

Damian-Nordic commented:
@wqx6 I moved your implementation in `OTAUpdater_esp32.h` to `platform/ESP32/OTAImageProcessorImpl.h` to make compliant with the new interface, but I don't have any ESP32 board to test it. Could you help me verify that everything works fine and append a fix to this PR if needed?

Damian-Nordic commented:
I'm not sure how to workaround the problem mentioned here: https://github.com/project-chip/connectedhomeip/pull/12722. Hence the build errors. Ideally, all apps that have OTA requestor server enabled have also OTA provider client enabled, but that doesn't seem to work.

dhrishi commented:
>@wqx6 I moved your implementation in OTAUpdater_esp32.h to platform/ESP32/OTAImageProcessorImpl.h to make it compliant with the new interface, but I don't have any ESP32 board to test it. Could you help me verify that everything works fine and append a fix to this PR if needed?

@shubhamdp @wqx6 PTAL

shubhamdp reviewed (commented):
@Damian-Nordic I tried building for ESP32 target but looks like build is failing. I have added few suggestions that I came across while building the example. Also, this PR may require a rebase as well. So I'll try after that as well.

Damian-Nordic commented on src/lib/shell/commands/BUILD.gn:
```diff
@@ -53,17 +53,10 @@ source_set("commands") {
   if (chip_enable_ota_requestor && chip_device_platform != "none") {
     sources += [ "Ota.cpp" ]
-    configs += [ "${chip_root}/src/controller:config" ]
+    public_deps += [ "${chip_root}/src/app/clusters/ota-requestor:headers" ]
```
Right, although I don't understand why. It works for other platforms.

Damian-Nordic commented on examples/ota-requestor-app/linux/LinuxOTAImageProcessor.cpp:
```diff
@@ -40,6 +40,11 @@ CHIP_ERROR LinuxOTAImageProcessor::Finalize()
     return CHIP_NO_ERROR;
```
I don't know if on Linux it is possible to have any reference implementation. The current one seems pretty "dummy" I would say.

Damian-Nordic commented on src/lib/shell/commands/Ota.cpp:
```diff
@@ -15,377 +15,84 @@
-}
-
-void OnConnectionFailure(void * /* context */, NodeId nodeId, CHIP_ERROR error)
+void InitRequestor()
```
Yeah, I thought about that, however there is not a single main.cpp, so I thought that maybe we could initialize the OTA Requestor in the shell until all examples implement OTA requestor.

Damian-Nordic replied:
Actually, I'll do as you suggest.

Damian-Nordic commented on src/lib/shell/commands/Ota.cpp:
```diff
@@ -411,12 +118,9 @@ void RegisterOtaCommands()
-        { &QueryImageHandler, "query",
-          "Query for a new image. Usage: ota query <fabric-index> <provider-node-id> <endpoint-id> <ip-address> <port>" },
-        { &ShowUpdateHandler, "show-update", "Print the current update information" },
+        { &QueryImageHandler, "query", "Query for a new image. Usage: ota query <fabric-index> <provider-node-id> <endpoint-id>" },
```
I think spec does not forbid OTA requestor to query for images at any time (except it shouldn't happen more often than once per 24 hours, so I would prefer having a way to trigger the query more explicitly.

Damian-Nordic commented:
@shubhamdp I removed OTA Image Processor for ESP32 as it's being added in https://github.com/project-chip/connectedhomeip/pull/13017. Also, I reworked the shell commands to use `GetRequestorInterface()` instead of creating the OTA Requestor instance in the shell command itself. Now it's up to the application to initialize the requestor properly and only then shell commands will work.
